src/Ast.ml

```diff
@@ -116,6 +116,194 @@ let rec is_sugared_list exp =
       is_sugared_list tl
   | _ -> false
 
+let doc_atrs atrs =
+  let doc, rev_atrs =
+    List.fold atrs ~init:(None, []) ~f:(fun (doc, rev_atrs) atr ->
+        let open Asttypes in
+        match (doc, atr) with
+        | ( None
+          , ( { txt= ("ocaml.doc" | "ocaml.text") as txt
+              ; loc= {loc_ghost= true} }
+            , PStr
+                [ { pstr_desc=
+                      Pstr_eval
+                        ( { pexp_desc=
+                              Pexp_constant (Pconst_string (doc, None))
+                          ; pexp_loc= loc
+                          ; pexp_attributes= [] }
+                        , [] ) } ] ) ) ->
+            (Some ({txt= doc; loc}, String.equal "ocaml.text" txt), rev_atrs)
+        | _ -> (doc, atr :: rev_atrs) )
+  in
+  (doc, List.rev rev_atrs)
+
+module type Module_item = sig
+  type t
+
+  val has_doc : t -> bool
+
+  val is_simple : t * Conf.t -> bool
+
+  val allow_adjacent : t * Conf.t -> t * Conf.t -> bool
+end
+
+module Structure_item : Module_item with type t = structure_item = struct
+  type t = structure_item
+
+  let has_doc itm =
+    match itm.pstr_desc with
+    | Pstr_attribute atr -> Option.is_some (fst (doc_atrs [atr]))
+    | Pstr_eval (_, atrs)
+     |Pstr_value (_, {pvb_attributes= atrs} :: _)
+     |Pstr_primitive {pval_attributes= atrs}
+     |Pstr_type (_, {ptype_attributes= atrs} :: _)
+     |Pstr_typext {ptyext_attributes= atrs}
+     |Pstr_exception {pext_attributes= atrs}
+     |Pstr_recmodule ({pmb_expr= {pmod_attributes= atrs}} :: _)
+     |Pstr_modtype {pmtd_attributes= atrs}
+     |Pstr_open {popen_attributes= atrs}
+     |Pstr_extension (_, atrs)
+     |Pstr_class_type ({pci_attributes= atrs} :: _)
+     |Pstr_class ({pci_attributes= atrs} :: _) ->
+        Option.is_some (fst (doc_atrs atrs))
+    | Pstr_include
+        {pincl_mod= {pmod_attributes= atrs1}; pincl_attributes= atrs2}
+     |Pstr_module {pmb_attributes= atrs1; pmb_expr= {pmod_attributes= atrs2}}
+      ->
+        Option.is_some (fst (doc_atrs (List.append atrs1 atrs2)))
+    | Pstr_value (_, [])
+     |Pstr_type (_, [])
+     |Pstr_recmodule []
+     |Pstr_class_type []
+     |Pstr_class [] ->
+        false
+
+  let rec is_simple_mod me =
+    match me.pmod_desc with
+    | Pmod_apply (me1, me2) -> is_simple_mod me1 && is_simple_mod me2
+    | Pmod_functor (_, _, me) -> is_simple_mod me
+    | Pmod_ident _ -> true
+    | _ -> false
+
+  let is_simple (itm, c) =
+    match c.Conf.module_item_spacing with
+    | `Compact ->
+        Location.width itm.pstr_loc <= c.Conf.margin
+        && Location.is_single_line itm.pstr_loc
+    | `Sparse -> (
+      match itm.pstr_desc with
+      | Pstr_include {pincl_mod= me} | Pstr_module {pmb_expr= me} ->
+          is_simple_mod me
+      | Pstr_open _ -> true
+      | _ -> false )
+
+  let allow_adjacent (itmI, cI) (itmJ, cJ) =
```

Comment:
It is not a big deal, but note that for my comment about making it easier to keep structures and signatures in sync, the idea was to have _one_ function that handled them both, to make it clearer that conceptually they are one thing.

Also, note that while I don't have any issue with this module style, the rest of the existing code that needs to operate over different types of phrase is all expressed in terms of Ast.t.